compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala

```diff
@@ -0,0 +1,219 @@
+import dotty.tools.dotc.core.quoted._
+import dotty.tools.dotc.core.NameKinds._
+import dotty.tools.dotc.core.StagingContext._
+import dotty.tools.dotc.core.StdNames._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.tasty.TreePickler.Hole
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.Spans._
+import dotty.tools.dotc.transform.SymUtils._
+import dotty.tools.dotc.transform.TreeMapWithStages._
+import dotty.tools.dotc.typer.Implicits.SearchFailureType
+import dotty.tools.dotc.typer.Inliner
+
+import scala.collection.mutable
+import dotty.tools.dotc.util.SourcePosition
+
+import scala.annotation.constructorOnly
+
+/** Checks that the Phase Consistency Principle (PCP) holds and heals types.
+ *
+ *  Type healing consists in transforming a phase inconsistent type `T` into `implicitly[Type[T]].unary_~`.
+ */
+class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(ictx) {
+  import tpd._
+
+  override def transform(tree: Tree)(implicit ctx: Context): Tree = tree match {
+    case tree: DefDef if tree.symbol.is(Inline) && level > 0 => EmptyTree
+    case _ => checkLevel(super.transform(tree))
+  }
+
+  /** Transform quoted trees while maintaining phase correctness */
+  override protected def transformQuotation(body: Tree, quote: Tree)(implicit ctx: Context): Tree = {
+    val body1 = transform(body)(quoteContext)
+    super.transformQuotation(body1, quote)
+  }
+
+  /** Transform splice
+   *  - If inside a quote, transform the contents of the splice.
+   *  - If inside inlined code, expand the macro code.
+   *  - If inside of a macro definition, check the validity of the macro.
+   */
+  protected def transformSplice(splice: Select)(implicit ctx: Context): Tree = {
+    if (level >= 1) {
+      val body1 = transform(splice.qualifier)(spliceContext)
+      val splice1 = cpy.Select(splice)(body1, splice.name)
+      if (splice1.isType) splice1
+      else addSpliceCast(splice1)
+    }
+    else {
+      assert(!enclosingInlineds.nonEmpty, "unexpanded macro")
+      assert(ctx.owner.isInlineMethod)
+      if (Splicer.canBeSpliced(splice.qualifier)) { // level 0 inside an inline definition
+        transform(splice.qualifier)(spliceContext) // Just check PCP
+        splice
+      }
+      else { // level 0 inside an inline definition
+        ctx.error(
+          "Malformed macro call. The contents of the $ must call a static method and arguments must be quoted or inline.",
+          splice.sourcePos)
+        splice
+      }
+    }
+  }
+
+
+  /** Add cast to force boundaries where T and ~t (an alias of T) are used to ensure PCP.
+   *  '{   ~(...: T)  }  -->  '{   ~(...: T).asInstanceOf[T]  } --> '{   ~(...: T).asInstanceOf[~t]  }
+   */
+  protected def addSpliceCast(tree: Tree)(implicit ctx: Context): Tree = {
+    val tp = checkType(tree.sourcePos).apply(tree.tpe.widenTermRefExpr)
+    tree.cast(tp).withSpan(tree.span)
```

Comment:
It does not always need to cast, but `ensureConforms` would not work here because we always cast to a type that is equal modulo spliced type aliases. 

A recording of the splice into something that is not the type alias `$splice` in `quoted.Type` could avoid this cast altogether. I intend to do this in a future PR, I will need it to hide internal encoding from the user and in #5297.